core/test-client/src/trait_tests.rs

```diff
@@ -0,0 +1,225 @@
+	// assert_eq!(
+	//	backend.blockchain().leaves().unwrap(),
+	//	vec![a5.hash(), b4.hash(), c3.hash()]);
+
+	// A1 -> D2
+	let mut builder = client.new_block_at(&BlockId::Hash(a1.hash())).unwrap();
+	// this push is required as otherwise D2 has the same hash as B2 and won't get imported
+	builder.push_transfer(Transfer {
+		from: Keyring::Alice.to_raw_public().into(),
+		to: Keyring::Ferdie.to_raw_public().into(),
+		amount: 1,
+		nonce: 0,
+	}).unwrap();
+	let d2 = builder.bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, d2.clone()).unwrap();
+	assert_eq!(
+		backend.blockchain().leaves().unwrap(),
+		vec![a5.hash(), b4.hash(), d2.hash()]);
+}
+
+
+pub fn test_blockchain_query_by_number_gets_canonical<B>(backend: Arc<B>) where
+	B: backend::LocalBackend<runtime::Block, Blake2Hasher, RlpCodec>,
+{
+	// block tree:
+	// G -> A1 -> A2 -> A3 -> A4 -> A5
+	//		A1 -> B2 -> B3 -> B4
+	//			  B2 -> C3
+	//		A1 -> D2
+	let client = ::new();
+
+	// G -> A1
+	let a1 = client.new_block().unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, a1.clone()).unwrap();
+
+	// A1 -> A2
+	let a2 = client.new_block_at(&BlockId::Hash(a1.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, a2.clone()).unwrap();
+
+	// A2 -> A3
+	let a3 = client.new_block_at(&BlockId::Hash(a2.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, a3.clone()).unwrap();
+
+	// A3 -> A4
+	let a4 = client.new_block_at(&BlockId::Hash(a3.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, a4.clone()).unwrap();
+
+	// A4 -> A5
+	let a5 = client.new_block_at(&BlockId::Hash(a4.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, a5.clone()).unwrap();
+
+	// A1 -> B2
+	let mut builder = client.new_block_at(&BlockId::Hash(a1.hash())).unwrap();
+	// this push is required as otherwise B2 has the same hash as A2 and won't get imported
+	builder.push_transfer(Transfer {
+		from: Keyring::Alice.to_raw_public().into(),
+		to: Keyring::Ferdie.to_raw_public().into(),
+		amount: 41,
+		nonce: 0,
+	}).unwrap();
+	let b2 = builder.bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, b2.clone()).unwrap();
+
+	// B2 -> B3
+	let b3 = client.new_block_at(&BlockId::Hash(b2.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, b3.clone()).unwrap();
+
+	// B3 -> B4
+	let b4 = client.new_block_at(&BlockId::Hash(b3.hash())).unwrap().bake().unwrap();
+	client.justify_and_import(BlockOrigin::Own, b4.clone()).unwrap();
+
+	// // B2 -> C3
+	let mut builder = client.new_block_at(&BlockId::Hash(b2.hash())).unwrap();
+	// this push is required as otherwise C3 has the same hash as B3 and won't get imported
+	// TODO [snd] fix that this fails with Error(ApplyExtinsicFailed(Stale)
+	// builder.push_transfer(Transfer {
+	//	from: Keyring::Alice.to_raw_public().into(),
+	//	to: Keyring::Ferdie.to_raw_public().into(),
+	//	amount: 1,
+	//	nonce: 0,
```

Comment:
`nonce: 1` should fix it AFAICT

Reply:
thanks. that fixed the `Error(ApplyExtinsicFailed(Stale)` issue